### PR TITLE
[dtensor][4/N] refactor dispatching logic and add propagator

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92126
* #92931
* #90735
* #90734
* __->__ #90733

This PR refactors the dispatching logic to make it more clean, and
isolate the sharding propagation logic out to a separate class.

This is so that we can implement more complicated propagation features
later.

Differential Revision: [D42876251](https://our.internmc.facebook.com/intern/diff/D42876251)

### CODE REVIEW

File: torch/distributed/_tensor/prop.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        """
+        Propagate the sharding for an operator given the op_schema.
+        """
+        op_key = str(op_overload)
+        sharding_prop_func = self.op_to_rules.get(op_key, None)
+
+        if sharding_prop_func is None:
+            # step 1. If there's not even one sharding rule
+            # implemented for the operator, we fall back to
+            # local tensor compute, this is wront currently
```
Comment (XilunWu) [inline torch/distributed/_tensor/prop.py:31]: typo: wront

File: torch/distributed/_tensor/prop.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            # we will change the behavior to reshard to full
+            # replicate and do the computatation
```
Comment (XilunWu) [inline torch/distributed/_tensor/prop.py:33]: typo: computatation

**Review by XilunWu — APPROVED** [review]
Great work! Thx for refactoring op dispatching and sharding propagation facilities. some typos to fix.

File: torch/distributed/_tensor/dispatch.py [hunk 3]
```diff
@@ -105,120 +105,71 @@ def _reshape_alias(
...
+) -> object:
+    # first we need to lift some private aten aliases to public calls
+    if op_call in _CURRENT_DECOMPOSITION_TABLE:
+        return _CURRENT_DECOMPOSITION_TABLE[op_call](*args, **kwargs)
+
+    # STEP 0. See if threre're user defined custom aten operator
+    # implementations. Custom operators take the highest priority
+    if custom_dispatch_ops is not None and str(op_call) in custom_dispatch_ops:
+        # dispatch to user defined custom distributed tensor ops
+        return custom_dispatch_ops[str(op_call)](*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_tensor/dispatch.py:123]: Yes we should deprecate this once we move all ops to use propagation rules

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42876251).

**Review by fduwjj — APPROVED** [review]
LGTM

**Pull request closed without merging**